pkg/apis/capkk/v1beta1/kkcluster_types.go

```diff
@@ -0,0 +1,287 @@
+	BootstrapReadyMessage string = "Bootstrap is ready."
+
+	// ClusterReadyCondition will check if cluster is ready.
+	ClusterReadyCondition clusterv1.ConditionType = "ClusterReadyCondition"
+	// WaitingCheckClusterReadyReason is one `Reason` of ClusterReadyCondition.
+	WaitingCheckClusterReadyReason string = "WaitingForClusterReady"
+	// WaitingCheckClusterReadyMessage is a specification `Message` of ClusterReadyCondition.
+	WaitingCheckClusterReadyMessage string = "Waiting for initial bootstrap to ready, add control-plane and worker nodes into cluster."
+	// ClusterNotReadyReason is one `Reason` of ClusterReadyCondition.
+	ClusterNotReadyReason string = "ClusterNotReady"
+	// ClusterReadyReason is one `Reason` of ClusterReadyCondition.
+	ClusterReadyReason string = "ClusterReady"
+	// ClusterReadyMessage is a specification `Message` of ClusterReadyCondition.
+	ClusterReadyMessage string = "Cluster is ready."
+)
+
+const (
+	// ClusterFinalizer allows ReconcileKKCluster to clean up KK resources associated with KKCluster before
+	// removing it from the apiserver.
+	ClusterFinalizer = "kkcluster.infrastructure.cluster.x-k8s.io"
+
+	// DefaultNodeSelectorMode is select nodes randomly.
+	DefaultNodeSelectorMode = RandomNodeSelectorMode
+
+	// DefaultControlPlaneGroupName defines default control plane group name of the kubernetes cluster.
+	DefaultControlPlaneGroupName = "kube_control_plane"
+
+	// DefaultWorkerGroupName defines default worker group name of the kubernetes cluster.
+	DefaultWorkerGroupName = "kube_worker"
+
+	// DefaultClusterGroupName defines default cluster group name of the kubernetes cluster.
+	DefaultClusterGroupName = "k8s_cluster"
+)
+
+// KKClusterSpec defines the desired state of KKCluster.
+type KKClusterSpec struct {
+	// Distribution represents the Kubernetes distribution type of the cluster.
+	Distribution string `json:"distribution,omitempty"`
+
+	// NodeSelectorMode is the select mode of the node selector.
+	// +optional
+	NodeSelectorMode NodeSelectorMode `json:"nodeSelectorMode,omitempty"`
+
+	// ControlPlaneGroupName is the group name of kubernetes control plane nodes.
+	// +optional
+	ControlPlaneGroupName string `json:"controlPlaneGroupName,omitempty"`
+
+	// WorkerGroupName is the group name of kubernetes worker nodes.
+	// +optional
+	WorkerGroupName string `json:"workerGroupName,omitempty"`
+
+	// ClusterGroupName is the group name of kubernetes cluster, which contains control plane group and worker group.
+	// +optional
+	ClusterGroupName string `json:"clusterGroupName,omitempty"`
+
+	// InventoryRef is the reference of Inventory.
+	InventoryRef *corev1.ObjectReference `json:"inventoryRef,omitempty"`
+
+	// ConfigRef is the reference of Config.
+	// +optional
+	ConfigRef *corev1.ObjectReference `json:"configRef,omitempty"`
+
+	// ConfigRef is the reference of Pipeline.
+	PipelineRef *corev1.ObjectReference `json:"pipelineRef,omitempty"`
```

Comment:
it batter set it in annotation or label. It is the pipeline that is created through kkcluster, not that kkcluster’s creation depends on the pipeline.